### ISSUE TITLE
each() is deprecated

### ISSUE THREAD

**Issue by jesseleite** [IcecaveStudios/parity#14]
Hey love the library, but we're dealing with `each()` errors now 😢

![screen shot 2018-04-09 at 3 56 15 pm](https://user-images.githubusercontent.com/5187394/38519673-9ef5427c-3c0e-11e8-9150-5759720cc348.png)

I assume [this will need to be refactored](https://github.com/IcecaveStudios/parity/blob/ccd81d359e94da09655edaf32c4a8a1e4e86105f/src/Comparator/DeepComparator.php#L108-L109)?

**Comment by JesseLeite** [comment]
Curious, if you don't use alternatively, do you have an alternative you would recommend?  Or you just don't have a need for this kind of thing?

**Comment by JesseLeite** [comment]
Hey again!  Just pulled repo down and tried to run tests, but you don't have phpunit in composer.json, even as dev dependency?

**Comment by jmalloc** [comment]
> Cool thanks for reply! Curious, if you don't use alternatively, do you have an alternative you would recommend? Or you just don't have a need for this kind of thing?

I'm not aware of any alternative, we simply haven't had the need as you say.

> Hey again! Just pulled repo down and tried to run tests, but you don't have phpunit in composer.json, even as dev dependency?

It's possible this package predates PHPUnit being available in composer!  We used a custom test-runner called [archer](https://github.com/IcecaveStudios/archer) which required PHPUnit to be in `$PATH`, though I'm reasonably certain *does* work with PHPUnit if it is pulled in via composer - so perhaps try adding PHPUnit (probably at v3) as a dev dependency and running `vendor/bin/archer`. 

Thanks for taking the time to have a look. Let me know if you have any luck with the tests running under Archer, and if not I'll take some time to modernise the project a little; vendor PHPUnit like a normal person, remove Archer, drop PHP 5 support, etc :)

**Comment by JesseLeite** [comment]
Thought I might dive in and see if I could help you modernize a bit.  Installed phpunit:

```
composer require --dev phpunit/phpunit=5.0
```

And ran archer, but got:

<img width="1396" alt="screen shot 2018-04-11 at 7 46 20 pm" src="https://user-images.githubusercontent.com/5187394/38648775-0b0cc0ee-3dc1-11e8-9d7d-068e70d32578.png">

I notice archer is an archived repo though as well, not sure if it's worth getting that to work?  I assume you don't use that anymore?

No rush on any of this btw!

**Comment by jmalloc** [comment]
> I notice archer is an archived repo though as well, not sure if it's worth getting that to work? I assume you don't use that anymore?

Yeah, definitely unused.  Since we wrote that originally a lot has changed with GitHub, etc -- for example we just rely on codecov.io to provide coverage reports now. 

I'm actually having trouble finding a good example of how the repos are normally set up these days with PHPUnit, but I believe it's a pretty standard approach.

`bootstrap.php` can go, it's an artefact of Archer. Instead we just put a `phpunit.xml` file in the project root. Take a look at the `phpunit.xml` file here https://github.com/eloquent/phony/blob/master/phpunit.xml for an example. Note it's bootstrap script just includes the composer autoloader.

The `test/src` folder can be added to the `autoload-dev` section in `composer.json`, as per https://github.com/eloquent/phony/blob/master/composer.json#L36

I don't *think* there's too much else required to get the builds working.

**Comment by jmalloc** [comment]
I should clarify that when I say unused, it appears Archer is still used in many of the repos in this org, but we definitely don't use it for any new projects.

**Comment by JesseLeite** [comment]
That's fair! 

>I'm actually having trouble finding a good example of how the repos are normally set up these days with PHPUnit, but I believe it's a pretty standard approach.

I think it's basically just to require phpunit as dev dependency, and include `phpunit.xml.dist` in the package's repo.  That way, you take load off users who might want to PR.  They can run `composer install` then `phpunit`, and they have a working test suite out-of-box, using a working version and config to get them started.  There's nothing wrong with globally installing phpunit, but setting some defaults allows PRs to focus on the actual issue instead of repo setup.

IMO, a good example of this is https://github.com/laravel/framework

What are your thoughts?

**Comment by jmalloc** [comment]
Yep, that sounds perfect.

**Comment by michalzielanski** [comment]
Updated compareArray() method:
```php
protected function compareArray(array $lhs, array $rhs, &$visitationContext)
{
    reset($lhs);
    reset($rhs);

    while (true) {
        $leftKey = key($lhs);
        $leftVal = current($lhs);
        next($lhs);
        $rightKey = key($rhs);
        $rightVal = current($rhs);
        next($rhs);

        if ($leftKey === null && $rightKey === null) {
            break;
        } elseif ($leftKey === null) {
            return -1;
        } elseif ($rightKey === null) {
            return +1;
        }

        $cmp = $this->compareValue($leftKey, $rightKey, $visitationContext);
        if ($cmp !== 0) {
            return $cmp;
        }

        $cmp = $this->compareValue($leftVal, $rightVal, $visitationContext);
        if ($cmp !== 0) {
            return $cmp;
        }
    }

    return 0;
}
```

**Comment by Lcfvs** [comment]
Are you planning commit this fix, please ?

**Comment by jmalloc** [comment]
I'll attempt to make some time for this today.

**Comment by Kcharle** [comment]
Would love to see this fix committed as well :)

**Comment by jmalloc** [comment]
Hi all, this is now fixed in v2.0.0. Please note that support for PHP <=7.0 has been *dropped*, which is why it's a new major version.

I apologise for how long it took me to get around to this; I know how frustrating it can be!
